### ISSUE TITLE
Lossless transcoding from JPEG

### ISSUE THREAD

**Issue by magicgoose** [google/pik#40]
Quote from the README:
> We are planning to keep the format 8x8 DCT based

Is one of the goals is to be able to losslessly re-pack JPEG images to PIK (keeping the actual data of 8x8 blocks intact), saving some space due to improved entropy coding?  
If that's true that's going to be great IMO, as there are a lot of images for which the original uncompressed versions cannot be found anymore, but they could still benefit from some size reduction without further losses (and ending up in a format that's going to be widely supported, unlike Lepton or arithmetic coded JPEGs).

**Comment by jyrkialakuijala** [comment]
Currently this claim/plan is vaporware -- no re-packing of jpegs is actually supported in pik. We have the technology, it is well tested, too, but we didn't yet figure out the correct way to deploy it -- and the time to do the deployment.

Our re-packer "brunsli" can recompress jpegs by 22 % (1 % less than Lepton) and uses 2.5x less cpu than Lepton in decoding. I am contemplating if such jpeg recompression would be better to be embedded into the content-encoding layer (like Brotli and Shared Brotli, possibly inside Shared Brotli) or if it should be inside an image codec. Sometimes JPEGs are inside a big file such as a tar or zip file (like Android APKs), and a lossless recompression there would help. Othertimes, they may be base64-encoded inside a web page. Having those two use cases covered would require that a more generic encoding (like the one in Shared Brotli) would improve those use cases.

On the other hand, having the image codec to deal with it might allow more interactive results or lower resource consumption in a browser viewing the images. Particularly, it feels more than a bit silly to reconstructing and repackiging a jpeg image from a recompressed image only for the purpose of viewing it a few milliseconds later.

Perhaps the correct solution is to add it in both compression use cases. What do you think?

**Comment by magicgoose** [comment]
I am mainly interested if the PIK **format** actually makes it possible to store 8x8 blocks from a JPEG as is, together with the correct colorspace, so that when decoded, this image looks exactly like the result of decoding the source JPEG file. And if not, then which changes to the format are necessary to make it a theoretically valid use case. **I've read that PIK uses its own colorspace which is nowhere near YUV, so then it will need to support YUV as well**. Perhaps something else is missing, too.  
Even if it's never going to be present in the official encoding tool (and that's totally fine), it will be made eventually **if the format supports it**. Writing another converter/encoder for an existing format is totally possible; but amending an already adopted specification could be practically impossible.  

> On the other hand, having the image codec to deal with it might allow more interactive results or lower resource consumption in a browser viewing the images. Particularly, it feels more than a bit silly to reconstruct and repackage a jpeg image from a recompressed image only for the purpose of viewing it a few milliseconds later.

Yes it is; and it won't work without javascript, and in a lot of other situations…
I saw that PIK format could be close to being a strict superset of JPEG (if it's going to use 8x8 DCT), so I decided to ask how feasible is it to alter it a little bit to solve this problem as well.

This all, of course, only makes sense if PIK format is going to eventually become as widely supported as JPEG is.
